src/shared/components/lazyMobXTable/LazyMobXTable.tsx

```diff
@@ -539,22 +560,26 @@ class LazyMobXTableStore<T> {
         return this.columnVisibility[column.name] || false;
     }
 
-    resolveColumnVisibility(columns:Array<Column<T>>): {[columnId: string]: boolean}
+    private resolveColumnVisibility(props: LazyMobXTableProps<T>): {[columnId: string]: boolean}
     {
-        const colVis:{[columnId: string]: boolean} = {};
-
-        columns.forEach((column:Column<T>) => {
-            // every column is visible by default unless it is flagged otherwise
-            let visible:boolean = true;
-
-            if (column.visible !== undefined) {
-                visible = column.visible;
-            }
+        let columnVisibility: {[columnId: string]: boolean};
 
-            colVis[column.name] = visible;
-        });
+        // if a custom columnVisibility object is provided use that one
+        if (props.columnVisibility) {
+            columnVisibility = props.columnVisibility;
+        }
+        // call the custom resolveColumnVisibility function if provided
+        // (this gives parent component full control over visibility)
+        else if (props.resolveColumnVisibility) {
+            columnVisibility = props.resolveColumnVisibility(props.columns, this._columnVisibility);
+        }
+        // if this.columnVisibility is already initialized, we are not trying to resolve it again by column definition
+        // this prevents inconsistencies between user selection and async programmatic updates
+        else  {
+            columnVisibility = this._columnVisibility ? this._columnVisibility : resolveColumnVisibility<T>(props.columns);
```

Comment:
Apparently this line breaks the Download tab screenshot test. Nice to see the tests in effect the day after we added them :)